### ISSUE TITLE
Support AVIF

### ISSUE THREAD

**Issue by GNURub** [imazen/imageflow#508]
I'd like to submit a proposal for a future feature to consider.

https://netflixtechblog.com/avif-for-next-generation-image-coding-b1d75675fe4

Format comparison:
https://jakearchibald.com/2020/avif-has-landed/

**Comment by lilith** [comment]
What's the binary size of the decoder and encoder?

**Comment by kornelski** [comment]
cavif is pretty lean. I've replaced libavif with basically one big `struct`. It's 100% pure Rust. Includes rav1e. Executable is 1.7MB, gzips to 500KB. 

`aom-decode` is fatter — 7MB — since it pulls in all of libaom. I haven't looked at trimming it down. Maybe dav1d would be smaller, but it needs meson and ninja to build, and I'm not a fan.

**Comment by lilith** [comment]
I'm hesitant to triple the size of Imageflow to add avif support.

**Comment by kornelski** [comment]
I get that decoding brings in large dependencies, but you could support encoding only. Encoding can be done in pure Rust and rav1e is pretty lean. I've released my encoder as a library: https://docs.rs/ravif

**Comment by lilith** [comment]
I will consider it then.

**Comment by iJungleboy** [comment]
Any progress on this? I would be very interested since AVIF support is at around 70% https://caniuse.com/avif and I believe it's really good.

**Comment by lilith** [comment]
I'm really hoping iOS adds support for AVIF. Safari gets much of its speed
by delegating image processing to the OS and its incredible hardware
integration, so stuff lands much more slowly there.

I would like to add this, but I have a lot going on right now, so I can't
offer a timeline.

**Comment by iJungleboy** [comment]
No problem - and thanks for your more-than-amazing work!

**Comment by stevetayloruk** [comment]
Support for this would be great with Safari soon to be supporting. https://www.coywolf.news/webmaster/safari-now-supports-avif-in-macos-ventura-and-ios-16/

**Comment by lilith** [comment]
It's planned. It will be slow, though, no getting around that.

**Comment by swiftMessenger** [comment]
AVIF support is around 93% now, and I've seen some adoption by the public as well. Is there an estimated timeline for when ImageFlow will support it? Also there was some (old) discussion about supporting an encoder but not a decoder. Is that still under consideration? The decoder is as important to me as the encoder.

**Comment by lilith** [comment]
It is still under consideration; it's orders of magnitude slower to encode,
however, so I'm currently focusing on bolstering caching and ahead-of-time
cache priming.

The decoder is also more security-critical surface area... and this format
will singlehandled double the binary size of the product and affect
scale-up times accordingly.

This week I'm adding aarch64 support for apple/win/linix.

**Comment by kornelski** [comment]
Proper caching is definitely required to compress AVIF properly. However, if you'd like *something* implementable sooner, then ravif/rav1e at the fastest encoding setting takes only about as long as encoding WebP, and is still a bit better than WebP.

**Comment by iJungleboy** [comment]
> aarch64

@tvatavuk this may be interesting for you as you were experimenting with ARM/Windows...

**Comment by lilith** [comment]
Supported aarch64 in source. But publishing is a problem since github
windows arm runners lack git, pwsh, bash, and everything.

**Comment by tvatavuk** [comment]
Thanks @lilith (and @iJungleboy) for making it possible to use Imageflow on Windows 11 ARM64. 😊 

I did a small proof of concept and manually built `imageflow.dll` on Windows 11 ARM64. Thanks to the great assets in your [Imageflow git repo](https://github.com/imazen/imageflow) and some help for a Rust newbie from Microsoft Copilot AI on [Windows Copilot+PC](https://www.asus.com/laptops/for-home/vivobook/asus-vivobook-s-15-s5507/), [2sxc](https://github.com/2sic/2sxc) is now showing images processed by Imageflow in [Oqtane](https://github.com/oqtane/oqtane.framework) also on Windows 11 ARM64 in IIS and IIS Express at least in my dev env. 🎉

Still, this is a build of the latest development version of `imageflow.dll` and is probably newer than the one used in 2sxc (I'm not sure about vertical compatibility). 🤔

This is a great step forward! 🚀

![20240830_173544](https://github.com/user-attachments/assets/5d6d19b7-ad75-405c-a3a6-61a52c92ba8a)

**Comment by lilith** [comment]
I'm glad to hear it! Sorry for the late response - I got arm64 published to nuget 3 months ago, but it's still such a pain on CI. 
https://www.nuget.org/packages/Imageflow.NativeRuntime.win-arm64/2.1.0-rc11

Back to the topic of AVIF, I'm thinking that I'd be better off porting jpegli than implementing avif. jpegli gives jpegs that often beat both webp and (high speed mode) avif files - and jpegli manages to be way faster than mozjpeg. And JXL is better than AVIF for all use cases as far as I can tell, with the exception of lower browser support (Safari and soon Firefox, but no news from Chrome yet). I think that will change once there's a mature rust decoder impl. 

https://cloudinary.com/blog/jpeg-xl-and-the-pareto-front

**Comment by DemiMarie** [comment]
Does AVIF have advantages when doing low-speed offline encoding?

**Comment by lilith** [comment]
Yes, although JXL is better, for lower quality images, with lots of CPU
time, avif is decent.

**Comment by iJungleboy** [comment]
@DemiMarie and Avif is currently supported by all the Apple products, while I believe JXL only has 14% or so coverage for now https://caniuse.com/jpegxl

**Comment by mark8055** [comment]
I see there is an AVIF branch, is this something that works and we could try running or is it still too far off?  AVIF seems to be taking off and I see it being used more then webp now in some contexts

**Comment by lilith** [comment]
I currently have 3 rust encoders I am working on with Fable; it has been my
main codec focus daily for a while now. I want to ship something
exceptional.
